pkg/prometheus/agent/statefulset.go

```diff
@@ -0,0 +1,478 @@
+				Protocol:      v1.ProtocolTCP,
+			},
+		}
+	}
+
+	volumes, promVolumeMounts, err := prompkg.BuildCommonVolumes(p, tlsAssetSecrets)
+	if err != nil {
+		return nil, err
+	}
+
+	// Mount web config and web TLS credentials as volumes.
+	// We always mount the web config file for versions greater than 2.24.0.
+	// With this we avoid redeploying prometheus when reconfiguring between
+	// HTTP and HTTPS and vice-versa.
+	webConfigGenerator := cg.WithMinimumVersion("2.24.0")
+	if webConfigGenerator.IsCompatible() {
+		var fields monitoringv1.WebConfigFileFields
+		if cpf.Web != nil {
+			fields = cpf.Web.WebConfigFileFields
+		}
+
+		webConfig, err := webconfig.New(prompkg.WebConfigDir, prompkg.WebConfigSecretName(p), fields)
+		if err != nil {
+			return nil, err
+		}
+
+		confArg, configVol, configMount, err := webConfig.GetMountParameters()
+		if err != nil {
+			return nil, err
+		}
+		promArgs = append(promArgs, confArg)
+		volumes = append(volumes, configVol...)
+		promVolumeMounts = append(promVolumeMounts, configMount...)
+	} else if cpf.Web != nil {
+		webConfigGenerator.Warn("web.config.file")
+	}
+
+	// The /-/ready handler returns OK only after the TSDB initialization has
+	// completed. The WAL replay can take a significant time for large setups
+	// hence we enable the startup probe with a generous failure threshold (15
+	// minutes) to ensure that the readiness probe only comes into effect once
+	// Prometheus is effectively ready.
+	// We don't want to use the /-/healthy handler here because it returns OK as
+	// soon as the web server is started (irrespective of the WAL replay).
+	readyProbeHandler := prompkg.ProbeHandler("/-/ready", cpf, webConfigGenerator, webRoutePrefix)
+	startupProbe := &v1.Probe{
+		ProbeHandler:     readyProbeHandler,
+		TimeoutSeconds:   prompkg.ProbeTimeoutSeconds,
+		PeriodSeconds:    15,
+		FailureThreshold: 60,
+	}
+
+	readinessProbe := &v1.Probe{
+		ProbeHandler:     readyProbeHandler,
+		TimeoutSeconds:   prompkg.ProbeTimeoutSeconds,
+		PeriodSeconds:    5,
+		FailureThreshold: 3,
+	}
+
+	livenessProbe := &v1.Probe{
+		ProbeHandler:     prompkg.ProbeHandler("/-/healthy", cpf, webConfigGenerator, webRoutePrefix),
+		TimeoutSeconds:   prompkg.ProbeTimeoutSeconds,
+		PeriodSeconds:    5,
+		FailureThreshold: 6,
+	}
+
+	podAnnotations, podLabels := prompkg.BuildPodMetadata(cpf, cg)
+	// In cases where an existing selector label is modified, or a new one is added, new sts cannot match existing pods.
+	// We should try to avoid removing such immutable fields whenever possible since doing
+	// so forces us to enter the 'recreate cycle' and can potentially lead to downtime.
+	// The requirement to make a change here should be carefully evaluated.
+	podSelectorLabels := map[string]string{
+		"app.kubernetes.io/name":        "prometheus",
+		"app.kubernetes.io/managed-by":  "prometheus-operator",
+		"app.kubernetes.io/instance":    promName,
+		"prometheus":                    promName,
```

Comment:
IIRC this label is legacy and we can probably get rid of it in favor of `app.kubernetes.io/instance`.

Reply:
I'd say only agent for now.